### DISCUSSION TITLE
issue: handleSubmit does not execute onSubmit function (or any other funciton) on react native

### DISCUSSION THREAD

**Discussion by massanen7** [react-hook-form/react-hook-form#8372] [Q&A]
### Version Number

7.31.1

### Codesandbox/Expo snack

https://snack.expo.dev/@bluebill1049/react-hook-form-v7---controller

### Steps to reproduce

1. Go to 'https://snack.expo.dev/@bluebill1049/react-hook-form-v7---controller'
2. Click on 'Button' button
3. See that console log of onSubmit function does not execute

### Expected behaviour

onSubmit function should execute, and you should see the console.log output.

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by Moshyfawn** [comment]
That's because you have field errors.
```json
{
  "errors": {
    "firstName": {
      "type": "required",
      "message": "",
      "ref": {
        "name": "firstName"
      }
    },
    "lastName": {
      "type": "required",
      "message": "",
      "ref": {
        "name": "lastName"
      }
    }
  }
}
```

if you want to do something, when the form submission fails, make sure to provide a second `onInvalid` callback to to the `handleSubmit` form method

```js
const onInvalid = (errors) => console.error(errors)

handleSubmit(onSubmit, onInvalid)
```

**Reply by pierratono** [reply]
onInvalid is very helful, saved my day

**Reply by mainAlik** [reply]
you save me with onInvalid 
thamk you <3

**Reply by MisterWaner** [reply]
Thank you so muuchhh !!!!! 
It's been 2 days that I'm stuck on a bug, and I find the solution here !

**Reply by SaddamMohsen** [reply]
thank you this is very helpfull

**Reply by Devlouiz** [reply]
Thanks alot 2 days of battling with this

**Comment by massanen7** [comment]
Nope, already tried that, and does not print anything. Anything I pass to the handleSubmit function, does not execute. I was first playing with my own project example, but now Im just playing in the example of the documentation, and happens exactly the same. Both onSubmit nor onInvalid, does not execute

**Comment by Moshyfawn** [comment]
Hm.. can't reproduce in [the Snack](https://snack.expo.dev/@moshyfawn/react-hook-form-v7---controller)

https://user-images.githubusercontent.com/16290753/168876788-8d8572d6-458c-489b-a4ac-ac6a222f14e3.mp4

**Comment by bluebill1049** [comment]
which example is in the doc? it's working for me: https://snack.expo.dev/@bluebill1049/react-hook-form-v7---controller

**Comment by Moshyfawn** [comment]
> Hm.. can't reproduce in [the Snack](https://snack.expo.dev/@moshyfawn/react-hook-form-v7---controller)
> 
> https://user-images.githubusercontent.com/16290753/168876788-8d8572d6-458c-489b-a4ac-ac6a222f14e3.mp4
> 
> 
> 

Hm, I see a couple of thumbs up on the issue. Can someone confirm it's not working in [this Snack](https://snack.expo.dev/@moshyfawn/react-hook-form-v7---controller)?

**Comment by massanen7** [comment]
https://user-images.githubusercontent.com/3900702/168992275-9d362016-6b90-4fcf-8efa-1cc5fbb9b118.mov

I'm working on typescript, but it all should work pretty much the same. Here is my recording. I also show that the reset actions works good, but not the handleSubmit one.

**Comment by bluebill1049** [comment]
it doesn't matter if it's typescript or not. please try to reproduce the issue in the snack above.

**Comment by massanen7** [comment]
Here you have pretty much the same structure I have. The "EmailInputScreen.tsx" is where I pretty much copy pasted the example of the documentation to play with react-hook-form. It's in a screen rendered in a stackNavigator, you'll see that onSubmit nor onErrors are never called there.

https://snack.expo.dev/@massanen/07c78c

**Reply by bluebill1049** [reply]
supply an isolated expo and I couldn't find `EmailInputScreen.tsx`
![Screen Shot 2022-05-18 at 7 35 21 pm](https://user-images.githubusercontent.com/10513364/169008345-bba6c64d-7747-4a90-b968-a0de82f449ef.png)

**Comment by bluebill1049** [comment]
update this codesandbox to reproduce the issue: https://snack.expo.dev/@bluebill1049/react-hook-form-v7---controller

**Reply by massanen7** [reply]
Well, I'm sorry but I can't update your codesandbox, when I do it will save in my own (which is the one I already shared). 

Here is the thing:
1. useForm works well when you try the example of the documentation in the root of the project.
2. When you use useForm in a screen component rendered by a StackNavigator the handleSubmit function from the useForm hook does not work anymore.

The same example of the documentation that we are talking about all the time: when it's in the root of the project, in App.js, it all works well. When it's in a Screen inside a navigation, it does not work. 

I might miss something for sure, but it's really really weird, and making me go nuts.

**Reply by bluebill1049** [reply]
I understand your frustration, switch side me and @Moshyfawn as the maintainers are here to try our best to help with our free time. If you don't share a straightforward example for us to reproduce, it doesn't help us either. here is 8 pm I am finishing off my work and i am tired and still trying to help but you need to give me a hand as well, so I can help you isolate the issue. 
![Screen Shot 2022-05-18 at 8 06 46 pm](https://user-images.githubusercontent.com/10513364/169014692-1d288393-431f-413f-8a5b-b943a931e3d2.png)

this snack you shared works: https://snack.expo.dev/@bluebill1049/react-hook-form-v7---controller
the snack that I shared works too.

now next is if you can isolate an snack then I can have a look at it again.

**Reply by massanen7** [reply]
Hey guys, I just found the issue:

The way I export my App.tsx. I had a condition depending on Config value to load either my App either Storybook.

This is what I had:
`export default Config.LOAD_STORYBOOK === 'true' ? StorybookUI : App;`

Just exporting the App without the condition with just:
`export default App;`

The useForm is working.

If this helps you or anybody in anyway I'm happy, but looks like it's more on my side. Thanks much for all the help, here is the video of how this change made me just fix it:


https://user-images.githubusercontent.com/3900702/169022305-6f65ca92-8124-43bc-a2d4-55fd401aef35.mov

**Reply by bluebill1049** [reply]
thanks for sharing. i am glad it finally works for you. good night.

**Reply by vinitsood** [reply]
Amazing! That was it. Thank you

**Reply by Rolozuna** [reply]
Thank you very much @massanen7, I was looking for the solution all morning, I searched everywhere, finally I found this answer, I was already thinking about using formik, haha. Thank you!

**Comment by Lokesh396** [comment]
I have a form in my component with some inputs and selects and i have some inputs which was not inside a form but i used register to get the data using **watch**, while submitting the form, the inputs which was not in the form is also restricting the form to submit, i got to know about the error using **onInvalid** is this the default behaviour of react-hook-form?

**Comment by anabeatrizzz** [comment]
What worked for me was to change function calling way

```typescript
 const onPressConfirm = (event: GestureResponderEvent) => {
  handleSubmit(onSubmit)(event);
};

<Button onPress={onPressConfirm}>
  Confirm
</Button>

```

Instead of `handleSubmit(onSubmit)`
